### PR TITLE
[MF-217] Record Patient Identifiers

### PR DESCRIPTION
**Pull request by samuelmale**
**Implemented objectives**
- Manually enter the identifier
- Support required VS non-required
- Dynamically load and display all identifier types*
- Support format validation (obtained from the back-end) 

---
*Ticket*: https://issues.openmrs.org/browse/MF-217

### CODE REVIEW

File: __mocks__/openmrs-esm-styleguide.mock.tsx
Comment (brandones) [inline __mocks__/openmrs-esm-styleguide.mock.tsx]: This kind of thing doesn't belong in a mock. Let tests do this themselves.

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [inline src/patient-registration/patient-registration.component.tsx]: It looks like the second call doesn't depend on the first. Since that's the case, let's [run them in parallel](https://stackoverflow.com/a/35612484/1464495).

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [inline src/patient-registration/patient-registration.component.tsx]: `pt` is short for "patient" (though outside medicine is also often "point" ). `pat` in the OpenMRS context is "Patient Attribute Type", so let's avoid that. But here you can just call this variable `identifiers`; there's no ambiguity as to what they correspond to.

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [inline src/patient-registration/patient-registration.component.tsx:184]: Nice! This seems like a big improvement. What sort of errors does this catch?

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [inline src/patient-registration/patient-registration.component.tsx]: It looks like something called `initialFormValues` was already being used on line 130/188 below. Where did `initialFormValues` come from before? Why did you introduce this new definition of it?

File: src/patient-registration/patient-registration.resource.tsx
Comment (brandones) [inline src/patient-registration/patient-registration.resource.tsx:37]: This is an interesting choice—why not just 
```
    }).then(response => response.data)
```

?

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [inline src/patient-registration/patient-registration.component.tsx]: Are identifiers demographic information? I think it would be nicer to just have
```jsx
  <DemographicsSection setFieldValue={props.setFieldValue} values={props.values} />
  <IdentifiersSection .......  />
  <ContactInfoSection />
```

File: src/patient-registration/patient-registration.resource.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.resource.tsx:37]: I intended to return an instance of a [PatientIdentifierType](https://github.com/samuelmale/openmrs-esm-patient-registration/blob/MF-217/src/patient-registration/patient-registration-helper.tsx#L34-L40). Seems I'm doing something superfluous due to my limited JavaScript knowledge?

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx]: Thanks for the StalkOverFlow pointer.

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [reply src/patient-registration/patient-registration.component.tsx]: But that's maybe a design decision— @nicholasjhill ?

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx:184]: @FlorianRappl explained to me back then that it depends on the browser's error API to handle any error that happened in the async call. See [thread](https://openmrs.slack.com/archives/CHP5QAE5R/p1594892146302400) in slack.

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx]: Nice catch @brandones . Seems this was a left over from a rebase against master.

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [reply src/patient-registration/patient-registration.component.tsx:184]: Yes, sorry, I was just curious whether there were examples of actual errors that you were seeing getting caught. No worries.

File: src/patient-registration/patient-registration.resource.tsx
Comment (brandones) [reply src/patient-registration/patient-registration.resource.tsx:37]: Oh sure, yeah. But the thing with the type-checking is that extra properties are allowed. You can treat `response.data` as a PatientIdentifierType object even if it has extra stuff on it. Does that make sense?

File: src/patient-registration/patient-registration.component.tsx
Comment (FlorianRappl) [reply src/patient-registration/patient-registration.component.tsx:184]: While the catch etc. is good the `createErrorHandler` should not be called *at all* somewhere. Maybe the implementation changed (then please forgive my comment), but in the implementation I know it should be treated like a singleton. `createErrorHandler` essentially couples to the browsers error handlers (e.g., onError) and thus would show multiple toasts if called (i.e., registered) multiple times.

It would make sense if the error handler would just register itself once and then would prevent further registrations.

As a result -> please remove the `createErrorHandler()`.

**Review by FlorianRappl — CHANGES_REQUESTED** [review]

File: src/patient-registration/patient-registration.resource.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.resource.tsx:37]: Yes it does, this means the JS engine will do the type casting for us.

Comment (samuelmale) [reply src/patient-registration/patient-registration.resource.tsx:37]: It would work but JS's smart Type casting may not infer [this property](https://github.com/samuelmale/openmrs-esm-patient-registration/blob/MF-217/src/patient-registration/patient-registration.resource.tsx#L34) for us automatically: 

```javascript

then(response => ({
      .......
      fieldName: camelCase(response.data.name),
      .......
    }));

```

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx:184]: Removed :)

File: __mocks__/openmrs-esm-styleguide.mock.tsx
Comment (samuelmale) [reply __mocks__/openmrs-esm-styleguide.mock.tsx]: Resolved

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx]: Resolved

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx]: Resolved

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx]: Resolved

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [reply src/patient-registration/patient-registration.component.tsx:184]: Nice catch, @FlorianRappl , it was about time I actually read the implementation of esm-error-handling :)

But just for future reference, `createErrorHandler` doesn't add the listener to `window.onError`. The listeners get added at the top level of `esm-error-handler`, so they can only happen once. What `createErrorHandler` does is (apparently, just reading [the code](https://github.com/openmrs/openmrs-esm-error-handling/blob/50053ebc76e0f9695f5731bc62b831be842d94c3/src/openmrs-esm-error-handling.ts#L29)) to ensure that the error contains a stack trace for the call site. So I guess its intended usage is like `.catch(createErrorHandler())`. Indeed, the way it had been used here wouldn't do anything.

File: src/patient-registration/patient-registration.resource.tsx
Comment (brandones) [reply src/patient-registration/patient-registration.resource.tsx:37]: Yeah, sort of—JS doesn't really do casting between objects at all. The object just gets passed through, and the receiving code tries to use it.

So yeah, you're absolutely right that the resulting object would not magically camelCase the `fieldName` property. You could camelCase at the time of usage, which might make the code simpler, and also make clearer _why_ the field is being camelCased in the first place. Or you could do

```javascript
    }).then(response => ({
      fieldName: camelCase(response.data.name),
      ...response.data
```

The only reason to write the code the way you've written it would be if the `response.data` object is many times larger than the data you intend to use. Which may be valid, but probably isn't worth worrying about right now.

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx:184]: @brandones It would actually, because it was in a `.catch(...)` block.

Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx:184]: I did it that way with something in my mind like: 

If there was an error in the request and the server responded with the custom error message, toast it like:

```java

showToast({ description: error.message });

```

If there was no custom error message, well lets call the default: `createErrorHandler()`. 

All the above wrapped in the `.catch()`

Comment (brandones) [reply src/patient-registration/patient-registration.component.tsx:184]: @samuelmale The following two blocks are different:

```javascript
.catch(createErrorHandler())
```

vs

```javascript
.catch(() => { createErrorHandler() })
```

the latter is more like what you've written. `createErrorHandler` is a function that returns a function that handles errors. You aren't doing anything with its return value.

Comment (brandones) [reply src/patient-registration/patient-registration.component.tsx:184]: Let's definitely add that `else` block back—we don't want to swallow any errors. But maybe we can write it like

```javascript
} else {
  createErrorHandler()(response)
}
```

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [reply src/patient-registration/patient-registration.component.tsx]: Ping @nicholasjhill  @nickjhill14 ? Or who's doing design on this, anyway?

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [reply src/patient-registration/patient-registration.component.tsx:184]: This is the only blocker before I hit the approve button. The other things would be nice but I'm not going to block on them.

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx]: @rbuisson also suggested to move the identifiers to their own section as you suggested above.

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [inline src/patient-registration/patient-registration.component.tsx]: As stated below, this is the only blocking issue. Let's try

```suggestion
          });
        } else {
          createErrorHandler()(response);
        }
```

Please verify that this works, at least by checking that the error is handled correctly when you block that URL:

Screenshot from 2020-08-25 14-16-51

File: src/patient-registration/patient-registration.component.tsx
Comment (brandones) [reply src/patient-registration/patient-registration.component.tsx]: Okay! Are you going to make that change now, or should we put this PR in as is?

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx:184]: Resolved

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx]: Done!

File: src/patient-registration/patient-registration.component.tsx
Comment (samuelmale) [reply src/patient-registration/patient-registration.component.tsx]: It works fine 👍

**Comment by samuelmale** [comment]
@brandones I have just rebased my working branch against master, I'm waiting for your approval

**Review by brandones — APPROVED** [review]

**Comment by brandones** [comment]
Thank you so much for your work on this, @samuelmale !

**Pull request merged**
